src/main/java/org/opensearch/security/securityconf/ConfigModelV7.java

```diff
@@ -753,35 +753,42 @@ public String getUnresolvedIndexPattern(User user) {
+        public Set<String> concreteIndexNames(final User user, final IndexNameExpressionResolver resolver, final ClusterService cs) {
+            return getResolvedIndexPattern(user, resolver, cs, false);
+        }
+
+        /** Finds the indices accessible to the user and attempts to resolve them to names, also includes any unresolved names */
+        public Set<String> attemptResolveIndexNames(final User user, final IndexNameExpressionResolver resolver, final ClusterService cs) {
+            return getResolvedIndexPattern(user, resolver, cs, true);
+        }
+
+        public Set<String> getResolvedIndexPattern(final User user, final IndexNameExpressionResolver resolver, final ClusterService cs, final boolean appendUnresolved) {
+            final String unresolved = getUnresolvedIndexPattern(user);
+            final ImmutableSet.Builder<String> resolvedIndices = new ImmutableSet.Builder<>();
+
+            final WildcardMatcher matcher = WildcardMatcher.from(unresolved);
             if (!(matcher instanceof WildcardMatcher.Exact)) {
                 final String[] aliasesForPermittedPattern = cs.state().getMetadata().getIndicesLookup().entrySet().stream()
                         .filter(e -> e.getValue().getType() == ALIAS)
                         .filter(e -> matcher.test(e.getKey()))
                         .map(e -> e.getKey())
                         .toArray(String[]::new);
-
                 if (aliasesForPermittedPattern.length > 0) {
-                    resolved = resolver.concreteIndexNames(cs.state(), IndicesOptions.lenientExpandOpen(), aliasesForPermittedPattern);
+                    final String[] resolvedAliases = resolver.concreteIndexNames(cs.state(), IndicesOptions.lenientExpandOpen(), aliasesForPermittedPattern);
+                    resolvedIndices.addAll(Arrays.asList(resolvedAliases));
                 }
             }
 
-            if (resolved == null && !unresolved.isEmpty()) {
-                resolved = resolver.concreteIndexNames(cs.state(), IndicesOptions.lenientExpandOpen(), unresolved);
+            if (Strings.isNotBlank(unresolved)) {
+                final String[] resolvedIndicesFromPattern = resolver.concreteIndexNames(cs.state(), IndicesOptions.lenientExpandOpen(), unresolved);
+                resolvedIndices.addAll(Arrays.asList(resolvedIndicesFromPattern));
             }
-            if (resolved == null || resolved.length == 0) {
-                return ImmutableSet.of(unresolved);
-            } else {
-                ImmutableSet.Builder<String> builder = ImmutableSet.<String>builder()
-                        .addAll(Arrays.asList(resolved));
-                if (appendUnresolved) {
-                    builder.add(unresolved);
-                }
-                return builder.build();
+
+            if (appendUnresolved || resolvedIndices.build().isEmpty()) {
+                resolvedIndices.add(unresolved);
```

Comment:
@peternied I've apporved the issue as this is not going to be an immidiate issue blocking this PR. But If there's no need to delay the return to line 791, it'd be safer to return the result at this line as what could happen is that when some indices are appended to `resolvedIndices` between `resolvedIndices.add(unresolved);` and `return resolvedIndices.build();` by accident, then those indices would unexpectedly included in the return result. Please consider changing this as a follow up.